### PR TITLE
consensus: Replay should check every block's apphash #3460

### PR DESCRIPTION
**Pull request by Zerppen**
Refs #3460

* [ ] Updated all relevant documentation in docs
* [ ] Updated all code comments where relevant
* [ ] Wrote tests
* [ ] Updated CHANGELOG_PENDING.md

### CODE REVIEW

**Comment by Liamsi** [comment]
Thanks again @Zerppen! In https://github.com/tendermint/tendermint/pull/3465#issuecomment-475271689 you asked why the tests fail:
the test `TestHandshakeReplayAll` combined with your changes:
```go
//check each block's apphash when replay
nextBlock := h.store.LoadBlock(i + 1)
```
fails. Note that the test uses the `mockBlockStore` which provides a mocked version of `LoadBlock`. Your changes call it with height 7 but there are only 6 blocks in the `chain` array (see the `mockBlockStore` struct).

**Comment by Liamsi** [comment]
Also, please format your code with `gofmt -s -w`.

**Comment by Zerppen** [comment]
yeah.It was my mistake

**Comment by melekes** [comment]
1) Why do we need to check block's app hash?
2) I don't think we need to load n+1 block. We can check the apphash here https://github.com/tendermint/tendermint/blob/25a3c8b1724c9611d6edc175b1b0d079f5ee28c1/consensus/replay.go#L419 

```go
	for i := appBlockHeight + 1; i <= finalBlock; i++ {
		h.logger.Info("Applying block", "height", i)
		block := h.store.LoadBlock(i)
		if appHash != nil {
                	checkBlockAppHash(block.AppHash, appHash)
		}
		appHash, err = sm.ExecCommitBlock(proxyApp.Consensus(), block, h.logger, state.LastValidators, h.stateDB)
		if err != nil {
			return nil, err
		}

		h.nBlocks++
	}
```

**Comment by Zerppen** [comment]
1.Current block's appash is made from last block.
2.And sm.ExecCommitBlock() return next block's apphash of app
3. but load n+1 block to get storeblock's apphash
Therefore, if first block's apphash of app returned is wrong followings must be wrong.
But current logic is only check last block's apphash.
@melekes

**Comment by melekes** [comment]
Okay. Sounds like checking app hash is a good idea 👍 

What about my 2nd point?

**Comment by Zerppen** [comment]
I think apphash was getting angry because you don't know he much.
To cheer apphash up,I gave you 3 points.
And I think reading each line of code can make you make sure he don't be angry.I mean apphash lol.

**Comment by Zerppen** [comment]
should be` h.store.LoadBlock(i + 1) `
not `h.store.LoadBlock( i )`

File: consensus/replay.go [hunk 1]
```diff
@@ -421,6 +421,18 @@ func (h *Handshaker) replayBlocks(state sm.State, proxyApp proxy.AppConns, appBl
 		if err != nil {
 			return nil, err
 		}
+		//check each block's apphash when replay
+		if !(h.store.Height() < i+1) {
+			nextBlock := h.store.LoadBlock(i + 1)
```
Comment (melekes) [inline consensus/replay.go]: I still think we can do this by saving prevBlock, not loading i+1 block

**Comment by melekes** [comment]
Closing in favor of https://github.com/tendermint/tendermint/pull/3579

**Pull request closed without merging**